### PR TITLE
Provide an option to reverse items only

### PR DESCRIPTION
**Pull request by knu**
I'm quite used to have a prompt at the bottom of the screen (not to mention vim & emacs) while I prefer a list of items shown from top down.  I just don't want to be distracted by what I type but just like to concentrate on what to pick.

So, I've extended the `--reverse` option so it optionally takes `all`, `items` or `none` so you can choose which area to reverse.

### CODE REVIEW

**Comment by junegunn** [comment]
I know people have different preferences and tastes, but I've tried to refrain from adding more options for tweaking UI elements for quite some time because as the maintainer of the project I have to draw the line somewhere to prevent fzf from becoming too complex. You can find my previous comment on that matter here: https://github.com/junegunn/fzf/issues/692#issuecomment-251436739

Having said that, I'm not particularly against the idea, but I must say that I'm not convinced either. But anyway, if I were to add the option, I would consider supporting all 4 possible layouts, with new `--layout=LAYOUT` option that supersedes `--reverse` like so:

- `--layout=bottom-up` (default)
- `--layout=top-down` (equivalent to `--reverse`)
- `--layout=bottom-up-below` (bottom-up layout, but prompt at the top. Not sure if this is a straightforward name for the option.)
- `--layout=top-down-above` (prompt at the bottom)

By the way, you might want to check if mouse click positions are properly calculated with your patch.

**Comment by knu** [comment]
Thanks for the feedback, I've fixed the mouse support.

Actually, I also considered the possibility of a --layout option, but I didn't feel the need for the fourth option: prompt at the top + items bottom up.  I believe the bottom-up list style is pretty uncommon; it only makes sense when the prompt is down below.

**Comment by junegunn** [comment]
Mouse pointer still isn't working on the prompt line.

Maybe you're right, `bottom-up-below` is probably not going to be the most popular option, then we can consider not implementing that particular layout. However, I still think it would be more straightforward to the users that we provide `--layout` option than to extend `--reverse` as suggested (we should keep `--reverse` for backward compatibility though). `--reverse` seemed like a good name 4 years ago when fzf was simple and small, but now fzf has so many options (including `--tac`) `--reverse` no longer feels like a straightforward name, and I'm concerned that making it do more things may bring extra confusion.

**Comment by knu** [comment]
I've addressed the mouse click issue and dropped the extension for `--reverse` in favor of a new `--layout` option.  What do you think?

File: man/man1/fzf.1 [hunk 1]
```diff
@@ -155,9 +155,22 @@ the full screen.
...
+.TP
+.BI "--layout=" "LAYOUT"
+Choose the layout (default: bottom-up)
+
+.br
+.BR bottom-up "     Show everything from bottom up"
+.br
+.BR top-down "      Show items from top down, prompt and header at the bottom"
+.br
+.BR top-down-above "Show everything from top down"
```
Comment (junegunn) [inline man/man1/fzf.1]: Well, I wanted `top-down` to be the new name for `--reverse`, and `top-down-above` to be your new layout. I mean "top-down-above (the prompt and header)". But I can definitely see that the term can be confusing, do you have a better idea?

File: src/options.go [hunk 2]
```diff
@@ -53,7 +53,9 @@ const usage = `usage: fzf [options]
                           height instead of using fullscreen
     --min-height=HEIGHT   Minimum height when --height is given in percent
                           (default: 10)
-    --reverse             Reverse orientation
+    --layout=LAYOUT       Choose layout: [bottom-up|top-down|top-down-above]
+                          (default: bottom-up)
+    --reverse             Deprecated synonym for --layout=top-down-above
```
Comment (junegunn) [inline src/options.go]: We can remove `--reverse` here as we still mention it on the man page.

File: src/options.go [hunk 3]
```diff
@@ -1037,10 +1062,13 @@ func parseOptions(opts *Options, allArgs []string) {
 			opts.Bold = true
 		case "--no-bold":
 			opts.Bold = false
+		case "--layout":
+			opts.Layout = parseLayout(
+				nextString(allArgs, &i, "layout required (bottom-up / top-down / top-down-above)"))
 		case "--reverse":
-			opts.Reverse = true
+			opts.Layout = layoutTopDownAbove
```
Comment (junegunn) [inline src/options.go]: So this becomes `layoutTopDown`.

File: man/man1/fzf.1 (hunk 1 above)
Comment (knu) [reply man/man1/fzf.1]: Ah, that's fine, I'll do that.

**Comment by junegunn** [comment]
Can you add a basic test case for your new layout? The testing code is a bit messy, but you'll get the hang of it.

https://github.com/junegunn/fzf/blob/master/test/test_go.rb

**Comment by knu** [comment]
@junegunn I think I've added just enough test cases, and while working on it I found and fixed a couple of bugs. 😆

**Comment by junegunn** [comment]
Cool, thanks! I'll test the feature and merge it in a few days.

**Comment by junegunn** [comment]
Looks like we had a miscommunication here, I commented above that I wanted `top-down-above`, as in "items are displayed in **top-down** layout, **above** the prompt". But I guess it equally makes sense to think "items are displayed in **top-down** layout, the prompt is shown **below** the list", so no, it's not really a good name 😕 

I'm open to suggestions. Some ideas:

- `upward`, `downward`, `downward/bottom` or `downward2`
- `asc`, `desc`, `desc/bottom`
- `1`, `2`, `3` (seriously)

**Comment by knu** [comment]
I see...  What about the following names, then?

- `upward` (default, aka `--no-reverse`)
- `downward-top` (new)
- `downward-bottom` (aka `--reverse`)

**Comment by junegunn** [comment]
I've thought about the options, but none of them really clicked with me. So I'd like to go with:

- `--layout=default`
- `--layout=reverse`
- `--layout=reverse-list`

These are quite similar to your initial proposal, but I still think `--layout=reverse` is more clear than just `--reverse=something`. And can you remove "deprecation" message for `--reverse` from the man page? Let's just leave `--reverse` as a shortcut option for `--layout=reverse`. Thanks.

**Comment by knu** [comment]
OK, I'll do that.

File: README.md [hunk 4]
```diff
@@ -228,7 +228,7 @@ You can add these options to `$FZF_DEFAULT_OPTS` so that they're applied by
 default. For example,
 
 ```sh
-export FZF_DEFAULT_OPTS='--height 40% --reverse --border'
+export FZF_DEFAULT_OPTS='--height 40% --layout=reverse-list --border'
```
Comment (junegunn) [inline README.md]: I want to present `--layout=reverse` as an example instead of the new `reverse-list`.

File: README.md [hunk 5]
```diff
@@ -266,7 +266,7 @@ or `py`.
     - e.g. `export FZF_DEFAULT_COMMAND='fd --type f'`
 - `FZF_DEFAULT_OPTS`
     - Default options
-    - e.g. `export FZF_DEFAULT_OPTS="--reverse --inline-info"`
+    - e.g. `export FZF_DEFAULT_OPTS="--layout=reverse-list --inline-info"`
```
Comment (junegunn) [inline README.md]: Same here

File: src/options.go [hunk 6]
```diff
@@ -90,7 +91,8 @@ const usage = `usage: fzf [options]
 
   Environment variables
     FZF_DEFAULT_COMMAND   Default command to use when input is tty
-    FZF_DEFAULT_OPTS      Default options (e.g. '--reverse --inline-info')
+    FZF_DEFAULT_OPTS      Default options
+                          (e.g. '--layout=reverse-list --inline-info')
```
Comment (junegunn) [inline src/options.go]: Here too.

**Comment by knu** [comment]
Done!

**Comment by junegunn** [comment]
Thanks, everything looks good now. I'll merge this just before I release 0.17.4, so that there's no confusion from the documentation on master branch vs. 0.17.3 binary.

**Comment by knu** [comment]
Great, thanks!

**Comment by junegunn** [comment]
Merged, thanks!

**Pull request merged**